metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+
+      scale_f = -np.sqrt(iter+1) / self.gamma
+
+      # proximal operator with negative trimming equivalent
+      w = scale_f * np.minimum(avg_grad_w + self.beta, 0)
+
+    if(self.verbose):
+      print("max iteration reached.")
+
+    # return L matrix yielded from best weights
+    self.components_ = self._get_components(best_w, basis)
+
+    return self
+
+  def _compute_dist_diff(self, T, X, basis):
+    """
+    Helper function to compute the distance difference of every triplet in the
+    space yielded by the basis set.
+    """
+    # Transformation of data by the basis set
+    XB = np.matmul(X, basis.T)
+
+    lenT = len(T)
+    # get all positive and negative pairs with lowest index first
+    # np.array (2*lenT,2)
+    T_pairs_sorted = np.sort(np.vstack((T[:, [0, 1]], T[:, [0, 2]])),
+                             kind='stable')
+    # calculate all unique pairs and their indices
+    uniqPairs, indices = np.unique(T_pairs_sorted, return_inverse=True,
+                                   axis=0)
+    # calculate L2 distance acording to bases only for unique pairs
+    dist = np.square(XB[uniqPairs[:, 0], :]-XB[uniqPairs[:, 1], :])
+
+    # return the diference of distances between all positive and negative
+    # pairs
+    return dist[indices[:lenT]]-dist[indices[lenT:]]
+
+  def _get_components(self, w, basis):
+    """
+    get components matrix (L) from computed mahalanobis matrix
+    """
+
+    # get rid of inactive bases
+    active_idx, = w > 0
+    w = w[..., active_idx]
+    basis = basis[active_idx, :]
+
+    K, d = basis.shape
+
+    if(K < d):  # if metric is low-rank
+      return np.sqrt(w.T)*basis  # equivalent to np.diag(np.sqrt(w)).dot(B)
+
+    else:   # if metric is full rank
+      return np.linalg.cholesky(np.matmul(basis.T, w.T*basis)).T
+
+  def _to_index_points(self, triplets, X=None):
+    shape = triplets.shape
+    X, triplets = np.unique(np.vstack(triplets), return_inverse=True, axis=0)
+    triplets = triplets.reshape(shape[:2])
+    return triplets, X
+
+  def _initialize_basis(self, triplets, X):
+    """ TODO: complete function description
+    """
+    n_features = X.shape[1]
+
+    # TODO:
+    # Add other options passed as string
+    authorized_basis = ['triplet_diffs']
+    if isinstance(self.basis, np.ndarray):
+      # TODO: should copy?
+      basis = check_array(self.basis, copy=True)
+      n_basis = basis.shape[0]
+      if basis.shape[1] != n_features:
+        raise ValueError('The input dimensionality ({}) of the given '
+                         'linear transformation `init` must match the '
+                         'dimensionality of the given inputs `X` ({}).'
+                         .format(basis.shape[1], n_features))
+    elif self.basis not in authorized_basis:
+      raise ValueError(
+          "`basis` must be '{}' "
+          "or a numpy array of shape (n_basis, n_features)."
+          .format("', '".join(authorized_basis)))
+    if self.basis is authorized_basis[0]:
```

Comment:
not great for readability

Reply:
Agreed, I was following what was done in the `init` handling on `_util.py`. I will improve it!

Reply:
Maybe I misunderstood you, you mean the message or the use of `authorized_basis[0]`?

Reply:
I meant the latter. `authorized_basis` is convenient to check if the value is allowed, but then when dealing with each case it is more readable to use the string values directly

Reply:
Agreed, already changed to that in the last commit!